### ISSUE TITLE
[wish] nuget proxy/cache

### ISSUE THREAD

**Issue by raffaeler** [NuGet/Home#458]
[Discussed via email with Yishai Galatzer]

In order to get fast access to nuget, provide a nuget proxy/cache component:
- installed on the local machine, it would allow to access packages very fast and in absence of network
- installed on the local network, it would avoid for large teams to download several times the same packages

Main differences with the local nuget repository (like a network share):
- the behavior should be transparent (cache) so that it is not necessary to copy the packages in the local repo 
- it should allow to accept a filter (proxy) so that certain sets of packages can't be accessed. This is mandatory for large companies where they must control the access to each nuget package. Most common filters:
* filter by author (for example all packages from MS are allowed)
* filter by license (for example all GPL packages must be blocked)
* filter by a specific version of a package (known to contains a bug, break compatibility, etc.)

**Comment by raffaeler** [comment]
Adding a comment now that VS2015 is RTM

In VS2015 when you try to manage the *installed* packages, the network is accessed.
Therefore if there is any issue with the network, you just stop working.
Please *detach* the network availability from all the nuget features. Network should be there only to list and download new packages, not the local/cached/installed ones.

**Comment by csharpfritz** [comment]
@raffaeler I think some of these items can be addressed with fixes in line with #885

**Comment by raffaeler** [comment]
Probably, but I consider the need of network a blocking issue.
Depending on the network is not acceptable at all.

**Comment by jshield** [comment]
I think ProGet (http://inedo.com/proget/overview) covers those requirements. You can aggregate feeds and filter them by license/company.

**Comment by andreabalducci** [comment]
totally agree @raffaeler networks is a blocking issue, +1 for  the lan cache service.
This morning we have issues with nuget.org (tracert show timeouts after msn network).
Results? Devs blocked and build server failing.
We should go to the beach instead of keep trying using nuget today.

**Comment by alkampfergit** [comment]
networks is always an issue with nuget, especially if you are behind a proxy or Authenticated Proxy. I've customer with slow network where using nuget is a real pain.

We really need a simple an easy way to at least setup a local nuget cache on local network that can manage restore packages without relying on the internet as @raffaeler suggested.

**Comment by raffaeler** [comment]
@jshield this topic is about a serious nuget issue and I personally do not accept posting advertising on expensive products.

**Comment by yishaigalatzer** [comment]
@raffaeler I think its fair to say that there is a commercial product cover this ask, and for large corps that want to block internet access from their employees that might even be a fair thing to setup. To be completely fair, the comment should include the fact that it is not _free_

@alkampfergit there is a way to set this up today with either nuget.server, nugetgallery or a host of other solutions (free or otherwise), but you have to download the packages you chose to make available and host them on that server.

For nuget.server (small amount of packages) - https://docs.nuget.org/create/hosting-your-own-nuget-feeds

https://github.com/NuGet/NuGetGallery/wiki/Hosting-the-NuGet-Gallery-Locally-in-IIS

This ask is for a more advanced proxy that allows accessing the live feed in a filtered/cached way.

**Comment by raffaeler** [comment]
@yishaigalatzer exactly, at least he should have specified it is not free. That is why I can only call it advertising.

**Comment by alkampfergit** [comment]
@yishaigalatzer I know, but it is a manual process. For small shop, people do not care about monitoring which packages should be available, they need a simple and easy way to have a local cache without user intervention. 

For big companies I'd like something that is capable of warning you if an "approved" packages is published on main nuget server and gives you the ability to decide if that update will be visible downstream etc.

**Comment by hvanbakel** [comment]
I'm testing this project, seems to be working just fine:
https://github.com/themotleyfool/Klondike

I'm using it as a local cache for internal packages. Works like a charm

**Comment by yishaigalatzer** [comment]
@hvanbakel coolness, I started looking into it over the weekend, but run out of free time. It is awesome that it works for you!

**Comment by jainaashish** [comment]
It has been an old and long discussion. And we don't think we'll get to this in near future since our focus has been on other items. So closing it as ByDesign
